salt/modules/boto_s3.py

```diff
@@ -0,0 +1,176 @@
+
+        s3.region: us-east-1
+
+    If a region is not specified, the default is us-east-1.
+
+    It's also possible to specify key, keyid and region via a profile, either
+    as a passed in dict, or as a string to pull from pillars or minion config:
+
+    .. code-block:: yaml
+
+        myprofile:
+            keyid: GKTADJGHEIQSXMKKRBJ08H
+            key: askdjghsdfjkghWupUjasdflkdfklgjsdfjajkghs
+            region: us-east-1
+
+:depends: boto3
+'''
+# keep lint from choking on _get_conn and _cache_id
+# pylint: disable=E0602
+
+# Import Python libs
+from __future__ import absolute_import
+import logging
+
+# Import Salt libs
+from salt.utils.versions import LooseVersion as _LooseVersion
+
+log = logging.getLogger(__name__)
+
+# pylint: disable=import-error
+try:
+    # pylint: disable=unused-import
+    import boto3
+    # pylint: enable=unused-import
+    import botocore
+    logging.getLogger('boto3').setLevel(logging.CRITICAL)
+    HAS_BOTO = True
+except ImportError:
+    HAS_BOTO = False
+# pylint: enable=import-error
+
+
+def __virtual__():
+    '''
+    Only load if boto libraries exist and if boto libraries are greater than
+    a given version.
+    '''
+    needed_boto3_version = '1.2.1'
+    msg = 'The boto_s3 module cannot be loaded: {0}.'
+    if not HAS_BOTO:
+        return (False, msg.format('boto3 libraries not found'))
+    if _LooseVersion(boto3.__version__) < _LooseVersion(needed_boto3_version):
+        submsg = 'boto3 library version {0} is required'.format(
+            needed_boto3_version,
+        )
+        return (False, msg.format(submsg))
+    return True
+
+
+def __init__(opts):  # pylint: disable=unused-argument
+    if HAS_BOTO:
+        __utils__['boto3.assign_funcs'](__name__, 's3')
+
+
+def get_object_metadata(
+    name,
+    extra_args=None,
+    region=None,
+    key=None,
+    keyid=None,
+    profile=None,
+):
+    '''
+    Get metadata about an S3 object.
+    Returns None if the object does not exist.
+
+    You can pass AWS SSE-C related args and/or RequestPayer in extra_args.
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt myminion boto_s3.get_object_metadata \\
+                         my_bucket/path/to/object \\
+                         region=us-east-1 \\
+                         key=key \\
+                         keyid=keyid \\
+                         profile=profile \\
+    '''
+    bucket, _, s3_key = name.partition('/')
+    if extra_args is None:
+        extra_args = {}
+
+    conn = _get_conn(region=region, key=key, keyid=keyid, profile=profile)
+
+    try:
+        metadata = conn.head_object(
+            Bucket=bucket,
+            Key=s3_key,
+            **extra_args
+        )
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Message'] == 'Not Found':
+            return {'result': None}
+        return {'error': __utils__['boto3.get_error'](e)}
+
+    return {'result': metadata}
+
+
+def upload_file(
```

Comment:
I think if you look at some of the earlier boto2 modules which Ryan and team developed, you'll see an initial approach of using `name` as the exclusive source for the primary state args, followed by a back-step where `name` is split off to allow for state names of the form `Ensure a file is in S3`, and adding explicit params for the state arg(s).  In this concrete case, I would request that `name` be used as a FALL-BACK source for `Bucket` and `Key` if the latter are not explicitly provided but separate params for the latter be added to allow arbitrary (descriptive rather than syntactic) `name` strings.  I gave an example of how to do this earlier in my code snippets.  Those should be CamelCase for ease of use, and I would think that ExtraArgs should also be camelcase, since it IS an explicit param which is sent to `upload_file()`.  That is, given a signature of
```
upload_file(Filename, Bucket, Key, ExtraArgs=None, Callback=None, Config=None)
```
`FileName`, `Bucket`, `Key`, `ExtraArgs` (and `Callback` and `Config` if they're ever added at a later date) should all be mapped directly to their CamelCase names, rather than munged into snake_case.  This holds true whether one is internally munging some subset of the data inside of ExtraArgs or not.  Also, renaming `FileName` to `source` is non-optimal for the same reasons - please leave it as `FileName` for ease of reference.  Ideally, `Callback` and `Config` would be included in the initial function, with doc strings mentioning they are not yet implemented (if simply passing through a valid value is not enough to make them work).

This should be done for both the execution module functions, AND the state module functions, wherever a param is mapped to one in the execution module function being called inside the state function.

So what I'm hoping to see -- if simply using `**kwargs` is not an option -- is all function params for `upload_file()` to be explicitly defined in the execution module function signature and have the same CamelCase param names, and that any args in the state function which map to execution function names do the same.  That is, the function signature in the exec module should be
```
upload_file(name, Filename, Bucket=None, Key=None, ExtraArgs=None, region=None, key=None, keyid=None, profile=None)
```
Then ensure that `Bucket` and `Key` are only parsed from `name` if `Bucket` and `Key` are not provided on the commandline.

The goal of all this is to 1) make the upstream documentation of AWS and boto3 as DIRECTLY APPLICABLE and USEFUL as possible, and 2) make the job of future maintainers easier by reducing the amount of "guaranteed overhead" they have to take on just to add new functionality.  The more we can map things directly, and SIMPLY without lots of "touching", the easier people's live will be later, both end-users and maintainers.

Reply:
I'm OK with using CamelCase args for the actual upload options (e.g. `ServerSideEncryption`), but I think following Salt conventions with snake_case for top level arguments looks nicer and is easier to follow.
I've used `source` instead of `FileName` to match the `file` states (e.g. `file.managed`) among others.

I think the benefits of a little sugar here for consistency with Python style and other Salt APIs outweigh those of directly shelling out to boto3. Trying to map all of the possible arguments is futile, but considering there's only ~3 top level arguments right now (name, source, and extra_args), I don't think this adds a significant maintenance burden.

Reply:
The major gist of my requests is, in fact, to MAP things as directly to the boto3 api as possible.  For reasons previously outlined, this makes maintainability, and future extension, as WELL as end-user ease of use, MUCH better.  The REASON snake_case is the norm for the existing boto2 based modules is because that's what boto2 DID, and thus the conventions of the time demanded mapping salt parameter names as directly as possible to the underlying boto2 params.  As previously mentioned, this is no longer the case, so the need is not for blindly following a given pattern when the rationale for it has changed, but rather to update the pattern when needed so the rationale which drives it is still served.

A side benefit is that you don't have to build extensive, and exhaustive, mapping tables of snake_case-to-CamelCase args, and update them every time something changes in boto3.  By utilizing a default pass-through approach, only munging params when AWS or boto force it due to brain-deadness on their part, you reap huge benefits in terms of maintainability and end-user convenience (not to mention the current development effort, which is greatly reduced, to boot).